Redox Support Preview

jackpot51 opened the pull request:
# Important - This is only a preview of a working `sys::redox`.

Compiling the Redox default distribution with this `libstd` results in a fully functioning distribution. As such, all further changes would be cosmetic or implementing features that have not been used by the default distribution (of which there are only a small number).

I do not expect this to be merged, but would like to discuss how it may be improved and get feedback.

There are a few `unimplemented!()` - `cloexec` for example. I have documented them below. These would be resolved before desiring a merge.

There are also issues with how the Redox syscall library is called - currently I am using a re-export in `libc` but that probably would not be desired.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @aturon (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

jackpot51 commented:
@alexcrichton @brson @eddyb I am mentioning you here because you have expressed interest in a Redox `libstd` port

Previous discussion here: https://internals.rust-lang.org/t/redox-support-in-liblibc-and-libstd/3954/17

jackpot51 commented on src/libstd/sys_common/mod.rs:
```diff
@@ -34,7 +34,6 @@ pub mod condvar;
 pub mod io;
 pub mod memchr;
 pub mod mutex;
-pub mod net;
```
Because Redox does networking using files, and does not support the traditional Berkeley Socket API, we must specify an alternative net module, which is defined in `sys::redox::net` and imported down below

jackpot51 commented on src/libstd/sys/redox/condvar.rs:
```diff
@@ -0,0 +1,97 @@
+    #[inline]
+    pub fn wait_timeout(&self, _mutex: &Mutex, _dur: Duration) -> bool {
+        ::sys_common::util::dumb_print(format_args!("condvar wait_timeout\n"));
+        unimplemented!();
```
This will be implemented shortly with additional flags to futex

jackpot51 commented on src/libstd/sys/redox/fd.rs:
```diff
@@ -0,0 +1,90 @@
+
+    pub fn set_cloexec(&self) -> io::Result<()> {
+        ::sys_common::util::dumb_print(format_args!("{}: set cloexec\n", self.fd));
+        unimplemented!();
```
This will be implemented shortly with an `fcntl` call

jackpot51 commented on src/libstd/sys/redox/fd.rs:
```diff
@@ -0,0 +1,90 @@
+
+    pub fn set_nonblocking(&self, _nonblocking: bool) -> io::Result<()> {
+        ::sys_common::util::dumb_print(format_args!("{}: set nonblocking\n", self.fd));
+        unimplemented!();
```
This will be implemented shortly with an `fcntl` call

jackpot51 commented on src/libstd/sys/redox/fs.rs:
```diff
@@ -0,0 +1,468 @@
+
+pub fn rename(_old: &Path, _new: &Path) -> io::Result<()> {
+    ::sys_common::util::dumb_print(format_args!("Rename\n"));
+    unimplemented!();
```
This will require a `rename` syscall

jackpot51 commented on src/libstd/sys/redox/fs.rs:
```diff
@@ -0,0 +1,468 @@
+
+pub fn set_perm(_p: &Path, _perm: FilePermissions) -> io::Result<()> {
+    ::sys_common::util::dumb_print(format_args!("Set perm\n"));
+    unimplemented!();
```
This will require a `chmod` syscall

jackpot51 commented on src/libstd/sys/redox/fs.rs:
```diff
@@ -0,0 +1,468 @@
+
+pub fn symlink(_src: &Path, _dst: &Path) -> io::Result<()> {
+    ::sys_common::util::dumb_print(format_args!("Symlink\n"));
+    unimplemented!();
```
Symlinks are not supported on Redox at the moment

nagisa replied:
In that case `Err` of some sort indicating that the operation is not supported should be returned (just like other OSes do when underlying filesystem does not support the operation).

jackpot51 commented on src/libstd/sys/redox/fs.rs:
```diff
@@ -0,0 +1,468 @@
+
+pub fn link(_src: &Path, _dst: &Path) -> io::Result<()> {
+    ::sys_common::util::dumb_print(format_args!("Link\n"));
+    unimplemented!();
```
Hard links are not supported on Redox at the moment

jackpot51 commented on src/libstd/sys/redox/pipe.rs:
```diff
@@ -0,0 +1,106 @@
+             _p2: AnonPipe,
+             _v2: &mut Vec<u8>) -> io::Result<()> {
+    ::sys_common::util::dumb_print(format_args!("read2\n"));
+    unimplemented!();
```
Due to using `set_nonblocking`, which is not supported until `fcntl` is added, this is not implemented yet

jackpot51 commented on src/libpanic_abort/lib.rs:
```diff
@@ -28,7 +28,7 @@
 #![feature(panic_runtime)]
 #![cfg_attr(unix, feature(libc))]
-#![cfg_attr(windows, feature(core_intrinsics))]
+#![cfg_attr(any(target_os = "redox", windows), feature(core_intrinsics))]
```
Redox has no `libc` glue for abort, so using the intrinsic is required

jackpot51 commented on src/libpanic_unwind/lib.rs:
```diff
@@ -69,6 +69,7 @@ mod imp;
 
 // i686-pc-windows-gnu and all others
 #[cfg(any(all(unix, not(target_os = "emscripten")),
+          target_os = "redox",
```
Redox could potentially use `libunwind` for backtraces. Right now, panic_abort is preferred.

jackpot51 commented on src/libstd/fs.rs:
```diff
@@ -348,6 +348,16 @@ impl File {
+    /// implemented on other operating systems using readlink
+    #[cfg(target_os = "redox")]
+    #[stable(feature = "rust1", since = "1.14.0")]
+    pub fn path(&self) -> io::Result<PathBuf> {
```
This is bad practice, and I will move this into a `os::redox::FileExt`

jackpot51 commented on src/libstd/io/stdio.rs:
```diff
@@ -81,11 +81,17 @@ impl Read for StdinRaw {
+    #[cfg(not(target_os = "redox"))]
     fn flush(&mut self) -> io::Result<()> { Ok(()) }
+    #[cfg(target_os = "redox")]
+    fn flush(&mut self) -> io::Result<()> { self.0.flush() }
```
On Redox, the `stdio` needs to be flushed when in raw mode. This perhaps could be removed by adjusting programs that enter raw mode

jackpot51 commented on src/libstd/sys/mod.rs:
```diff
@@ -32,6 +32,10 @@
 
+#[cfg(target_os = "redox")]
+#[path = "redox/mod.rs"]
+mod imp;
```
As we have a lot of different code, we create a new target family and a new `sys::redox` module

jackpot51 commented:
I added some annotations to the code to identify why I changed shared files (ones that are common to all targets), and why I left in some `unimplemented!()`'s

nagisa commented on src/libstd/sys/redox/net/tcp.rs:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    pub fn shutdown(&self, _how: Shutdown) -> Result<()> {
+        Err(Error::new(ErrorKind::Other, "TcpStream::shutdown not implemented"))
```
Might make sense to introduce `NotSupported` variant for ErrorKind.

@alexcrichton @aturon how is the list of variants in ErrorKind determined? Are they added on-demand?

KostaCitrix replied:
Yes, I would have found `NotSupported` useful when trying to implement a `sendfile()` library. That call just does not exist on Windows, but I didn't want to panic. There was no good error to return.

nagisa commented on src/libstd/sys/redox/os_str.rs:
```diff
@@ -0,0 +1,119 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
```
Is reexporting the `sys/unix/os_str` here not an option?

eddyb commented:
One concern I have with the `std::sync` concurrency primitives: could support be added in https://github.com/Amanieu/parking_lot instead? Then, a `parking_lot` backend for `std::sync` would benefit everyone.

brson commented:
I'm really excited about reading this, but not sure yet when.

bors commented:
:umbrella: The latest upstream changes (presumably #37886) made this pull request unmergeable. Please resolve the merge conflicts.

brson commented:
Hi, @jackpot51.

Sorry for taking so long to get back to you. I was quite afraid of this, but it
looks great! There are only a few tricky bits that are going to require some
tough decisions. I'm very appreciative that you've managed to corrale the vast
bulk of the code into sys::redox - do whatever you want there and it doesn't
require much deliberation. And thanks for adding the play-by-play comments to
the diff to explain some of the rationale for things that are not in sys::redox.

The libpanic_abort changes are fine for now. Ideally we would not want
platform-specific code here I think, but aren't in a position to do anything
about it yet. Perhaps libpanic_abort can one day depend on the hypothetical PAL.

For StdoutRaw::flush and StderrRaw::flush, I think we should sink the cfg into
std::sys - give sys::stdio::Stdout a flush method and make it a no-op everywhere
but Redox.

The sys_common::net change to redirect the entire module to a redox
implementation is fine for now. The network abstraction in std I suspect will
change a lot. Assuming that all networking is BSD networking obviously isn't
portable. I'd expect in the future that sys_common::net will be extracted into a
helper crate/mod that Unix/Windows can use and others can ignore.

I'm undecided about `std::os::unix`. It's kinda odd that the redox target
doesn't define `cfg(unix)` but still defines `os::unix`. And I think it's
basically incorrect that the redox definitions don't support the full API
surface here. At the same time, the practical short-term impact is quite low,
since we'd expect the redox target to continue to evolve for a while yet. So I
think I'm ok having `os::unix` defined for now, with the understanding that it's
kinda wrong and may be changed in the future.

I think we can't add the `path` method to `File` just for redox. Maybe it's a
useful in general and doable on all platforms (I think this subject was debated
when stabilizing `File`), but that would have to go through another process. For
now I would suggest achieving the functionality with a combination of AsRawFd
and an out-of-tree extension method. Alternately, we might be able to put an
extension trait in `os::redox`, though it looks to me like there is no such
precedent right now - besides `os::unix` and `os::windows` all the other
OS-specific modules seem to contain the exact same things, which seem to be
mostly for internal use (in retrospect I'm extremely skeptical of all these
public os::solaris, etc modules - seems like a big, open-ended,
platform-specific wart).

So all that seems like pretty minor stuff to me. That leaves the big one:
`Prefix::Scheme`. I was afraid `Path` would bite us someday, but hoped it
wouldn't be for many years yet. This enum is not open for extension, so it's
very difficult to make this change as written. If we could assure ourselves that
essentially no existing code was doing an exhaustive match on this enum then we
could consider a breaking change here, adding an unstable `__NonExhaustive`
variant in the process to open it for further extension, but I imagine this enum
is actually matched in the wild.

It probably is worth doing the experiment to find out how breaking it is, and
this matter may end up being a longer-term design question, but we probably need
to come up with a short-term workaround as well. You might commandeer one of the
other variants to use for Redox schemes. To be forward compatible against a
future addition of `Prefix::Scheme` you might have downstream code use extension
methods to query the schemyness of prefixes instead of using variants directly.

I haven't looked closely at the `sys::redox` code, but as I mentioned earlier we
can be less vigilant there.

So other than a few things, this looks pretty great. I think the things to do are

- Remove `File::path` and implement that functionality through out-of-tree
  extensions
- Remove `Prefix::Scheme` and instead parse Redox schemes into e.g. `Prefix::Verbatim`
- Maybe do a study of exhaustive matches on `Prefix` in the wild to determine
  the feasibility of a breaking change

jackpot51 commented:
@brson I have done the first 3 items and converted `sys::redox` to directly use `redox_syscall` as we discussed

I had to make a change to the `Prefix` length calculation on `Prefix::Verbatim` to support Redox's format. It is minimal, and can be seen [here](https://github.com/rust-lang/rust/pull/37702/files#diff-120c5906285d41e976dc4176265d7cbaR190)

jackpot51 commented on src/libstd/io/stdio.rs:
```diff
@@ -81,11 +81,11 @@ impl Read for StdinRaw {
 impl Write for StdoutRaw {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> { self.0.write(buf) }
-    fn flush(&mut self) -> io::Result<()> { Ok(()) }
+    fn flush(&mut self) -> io::Result<()> { self.0.flush() }
```
In order to support manual terminal refreshes, Redox allows applications to access fsync on stdout and stderr

jackpot51 commented on src/libstd/path.rs:
```diff
@@ -185,6 +187,9 @@ impl<'a> Prefix<'a> {
         }
         match *self {
+            #[cfg(target_os = "redox")]
+            Verbatim(x) => 1 + os_str_len(x),
```
Due to having only an extra b':', we add 1 to the verbatim string length instead of 4

brson commented:
This patch looks excellent now. There are few changes outside of `sys::redox`.

This patch now makes the redox port depend on the redox `syscall` crate instead of libc. I don't see the addition of `redox_syscall` to Cargo.toml so I guess it is arranged to be available by some other means. It's something to be aware of but I think it's ok for now. At some point it will probably want to be represented in cargo.

The changes to `Prefix::len` to repurpose `Prefix::Verbatim` are unfortunate https://github.com/rust-lang/rust/pull/37702/commits/2ec21327f25e4646ce64a2e331bb4702cf96dc99#diff-120c5906285d41e976dc4176265d7cbaL193.

The problem is that we can't add a `Prefix` variant that Redox needs, so we're trying to have Redox use `Prefix::Verbatim` for Redox schemes, but the `len` calculation for `Verbatim` is different for Redox than Windows.

This also raises the issue that using `Verbatim` to mean something different means that one can't count on `Verbatim` being a Windows path. I'm not sure that matters since ISTM that `Path` only really makes sense on the local platform - it doesn't represent any type of path, it represents a path using the semantics of the local OS.

I still don't have a better idea for how to immediately solve the problem that Redox needs another type of Prefix, so I'm inclined to further delegate the `len` method to `sys`, introducing the assumption that `Prefix::Verbatim` can mean different things on different platforms. I think we should probably get a gut check from other @rust-lang/libs people though.

jackpot51 commented:
@brson I appreciate the review. I don't have to add redox_syscall to Cargo.toml because I am adding it with a custom Cargo.toml.

What happens now? Wait for further reviews?

frewsxcv commented on src/libstd/sys/redox/os.rs:
```diff
@@ -0,0 +1,203 @@
+            for line in string.lines() {
+                if let Some(equal_sign) = line.chars().position(|c| c == '=') {
+                    let name = line.chars().take(equal_sign).collect::<String>();
+                    let value = line.chars().skip(equal_sign+1).collect::<String>();
```
https://doc.rust-lang.org/std/primitive.str.html#method.split_at might help you here

jackpot51 replied:
Cool, thanks

jackpot51 replied:
I took care of it by using splitn

brson commented:
@jackpot51 Yes, I think we need feedback from other people about how to deal with `Prefix`. I am not confident in the current patch.

The best option I have is to commandeer the `Verbatim` variant, as in this patch, and sink all the length calculations into `sys`, with the tradeoffs I mentioned in the previous comment.

jackpot51 commented:
@brson Just let me know who to annoy if I need to annoy them :P

alexcrichton commented:
Is there any way we can punt this path question to a later point? It's explicitly documented right now as Windows-only so jerry-rigging it for another platform is unlikely to work out that great (as we're seeing). If we can sidestep this question entirely that would be best.

jackpot51 commented:
@alexcrichton If it is Windows only, shouldn't it be part of sys::windows? Allowing each platform to define it's own path model?

I can easily undo changes there, but then the Redox port will be **incorrect** - Redox paths do have a prefix component that has meaning

alexcrichton commented:
It probably should have been moved, yes, but that's a moot question now as it's been stable since Rust 1.0.

Could you elaborate on the incorrectness you're thinking will arise? If we dig into it there's perhaps a way to work around it (but that's just a guess).

jackpot51 commented:
@alexcrichton Take, for instance, terminal auto completion. With Prefix the way it is in this patch, the autocompleter can correctly fill in paths that start with `/` like `/bin/cat`, relative paths like `cat`, as well as scheme-absolute paths like `file:/bin/cat`.

Without this patch, filling in `file:/bin/cat` will look for a folder named `file:` in the current directory

alexcrichton commented:
Well, I suppose it was just an idea. If it's require to have a custom prefix for redox, I don't know how to solve that problem in a good way.

brson commented:
I think we're at an impasse for now on the prefix issue. Can we just remove prefixes from this patch to get the bulk of the work landed and circle back to the tough design question?

I'll go post a thread to irlo about expanding Prefix to raise awareness and see if anybody has bright ideas.

jackpot51 commented:
@brson Sure

jackpot51 commented:
@brson @alexcrichton The prefix handling has been removed. This must be readded in the future to avoid leaving in a potential security problem, but I would like to see a merge of Redox support before then as you have described.

brson commented:
@bors r+

Thanks for your enduring patience.

bors commented:
:pushpin: Commit c61baa0 has been approved by `brson`

<!-- @bors r=brson c61baa0fc7a85bd2bcce34aac05ed739261cf037 -->

jackpot51 commented:
Thanks @brson. I am going to work on adding the `x86_64-unknown-redox` target to `rustc` now.

jackpot51 commented:
@brson, the [tidy job failed](https://travis-ci.org/rust-lang/rust/jobs/183423425) due to an issue fixed by https://github.com/rust-lang/rust/pull/38340.

brson commented:
@bors r+

bors commented:
:pushpin: Commit 3e15dc1 has been approved by `brson`

<!-- @bors r=brson 3e15dc108c66891da04aa8c3f77162746fab4277 -->

eddyb commented:
@bors r=brson

bors commented:
:pushpin: Commit 3e15dc1 has been approved by `brson`

<!-- @bors r=brson 3e15dc108c66891da04aa8c3f77162746fab4277 -->

jackpot51 commented:
thanks @eddyb

bors commented:
:hourglass: Testing commit 3e15dc1 with merge cfa668f...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/11284), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/11347), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/2571), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/2390), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/4558), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/11289), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/11237), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/3216), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/11210), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/4439), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/6389), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/11330), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/11304), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/3249), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/2429), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/6432), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/3283), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/6395), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/5581), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/2593), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/6407), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/3272)
Approved by: brson
Pushing cfa668f3bf8156730b419deec51fb980af23b8bb to master...

ticki commented:
woohoo!
